src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ReliableChannelBinder.cs

```diff
@@ -0,0 +1,2231 @@
+                }
+            }
+
+            _synchronizer.UnblockWaiters();
+            this.OnShutdown();
+            this.OnAbort();
+
+            if (channel != null)
+            {
+                channel.Abort();
+            }
+
+            this.TransitionToClosed();
+        }
+
+        protected virtual void AddOutputHeaders(Message message)
+        {
+        }
+
+        private bool CloseCore(out TChannel channel)
+        {
+            channel = null;
+            bool abort = true;
+            bool abortChannel = false;
+
+            lock (ThisLock)
+            {
+                if ((_state == CommunicationState.Closing)
+                    || (_state == CommunicationState.Closed))
+                {
+                    return true;
+                }
+
+                if (_state == CommunicationState.Opened)
+                {
+                    _state = CommunicationState.Closing;
+                    channel = _synchronizer.StopSynchronizing(true);
+                    abort = false;
+
+                    if (!this.MustCloseChannel)
+                    {
+                        channel = null;
+                    }
+
+                    if (channel != null)
+                    {
+                        CommunicationState channelState = channel.State;
+
+                        if ((channelState == CommunicationState.Created)
+                            || (channelState == CommunicationState.Opening)
+                            || (channelState == CommunicationState.Faulted))
+                        {
+                            abortChannel = true;
+                        }
+                        else if ((channelState == CommunicationState.Closing)
+                            || (channelState == CommunicationState.Closed))
+                        {
+                            channel = null;
+                        }
+                    }
+                }
+            }
+
+            _synchronizer.UnblockWaiters();
+
+            if (abort)
+            {
+                this.Abort();
+                return true;
+            }
+            else
+            {
+                if (abortChannel)
+                {
+                    channel.Abort();
+                    channel = null;
+                }
+
+                return false;
+            }
+        }
+
+        public Task CloseAsync(TimeSpan timeout)
+        {
+            return this.CloseAsync(timeout, _defaultMaskingMode);
+        }
+
+        public async Task CloseAsync(TimeSpan timeout, MaskingMode maskingMode)
+        {
+            this.ThrowIfTimeoutNegative(timeout);
+            TimeoutHelper timeoutHelper = new TimeoutHelper(timeout);
+            TChannel channel;
+
+            if (CloseCore(out channel))
+            {
+                return;
+            }
+
+            try
+            {
+                this.OnShutdown();
+                await this.OnCloseAsync(timeoutHelper.RemainingTime());
```

Comment:
Assuming you don't have some sort of IL rewriter running, you probably want `.ConfigureAwait(false)` on all of your awaitables.

Reply:
We don't need to worry about it. Because WCF public async api's are composites of lots of other api's and we pretty much always need to go async because of actual real network calls, we took a different approach. At an early point in the public api call, we check if we are using the default sync context and default task scheduler, and if we're not, we force an artificial async continuation onto a threadpool thread using the default Task Scheduler. This means we never have the ambiguous situation of not knowing if we're in a situation where ConfigureAwait is needed.

---
In reply to: [274046096](https://github.com/dotnet/wcf/pull/3538#discussion_r274046096) [](ancestors = 274046096)